### PR TITLE
Support torch.device

### PR DESCRIPTION
**Pull request by kshitij12345**
Fixes: https://github.com/Lightning-AI/lightning-thunder/issues/327

This PR enables using `torch.device` in the jitted function. For this PR, we bake the device into the trace rather than creating a symbol for `torch.device`.

Example:
```python
import thunder
import torch

def foo(dev, idx, t):
    return torch.ones(1, device=torch.device(dev, idx)) + torch.ones(1, device=t.device) * 2

jfoo = thunder.jit(foo)
# jfoo = thunder.jit(foo, cache=thunder.CACHE_OPTIONS.SYMBOLIC_VALUES)
print(jfoo("cuda", 0, torch.randn(3, device='cuda')))

# print(thunder.last_prologue_traces(jfoo)[-1])
print(thunder.last_traces(jfoo)[-1])

```

Trace
```
# Constructed by Delete Last Used (took 0 milliseconds)
import torch
from thunder.executors.torchex import no_autocast

@torch.no_grad()
@no_autocast
def computation():
  [t5] = nvFusion0()
    # t2 = prims.full((1,), 1, device=devices.Device("cuda:0"), dtype=dtypes.int64_)  # t2: "cuda:0 i64[1]"
    # result = prims.mul(t2, 2)  # result: "cuda:0 i64[1]"
    # t5 = prims.add(t2, result)  # t5: "cuda:0 i64[1]"
  return t5
```

--------

**Supporting `SYMBOLIC_VALUES` (in future)** - 
Above approach of baking device into the trace works fine with the current caching scheme but this will lead to issue with SYMBOLIC_VALUES caching mechanism (below snippet will fail with `SYMBOLIC_VALUES`). I think one way to support this will be to make `torch.device` actually a symbol which can take `int` or `IntegerProxy` and return a device proxy for consumption. However, I am not sure if this is the best approach. @jjsjann123 would love to know your thoughts and suggestions.

```python
import thunder
import torch

def foo(dev, idx):
    return torch.ones(1, device=torch.device(dev, idx))

# jfoo = thunder.jit(foo)  # works
jfoo = thunder.jit(foo, cache=thunder.CACHE_OPTIONS.SYMBOLIC_VALUES)
print(jfoo("cuda", 0))  # fails as our current implementation can't handle `IntegerProxy`.
print(jfoo("cuda", 1))

# print(thunder.last_prologue_traces(jfoo)[-1])
# print(thunder.last_traces(jfoo)[-1])
```

--------

**Testing** - 

Since, `torch.device` is not aTensor Operation (i.e. a function that takes or returns a Tensor) and also since this is not implemented as a Symbol in trace, I haven't used the OpInfo mechanism but instead written the tests in `test_core.py`.

### CODE REVIEW

**Comment by jjsjann123** [comment]
I'm interesting in seeing how the trace is constructed with `symbolic values`

> I think one way to support this will be to make torch.device actually a symbol which can take int or IntegerProxy and return a device proxy for consumption. However, I am not sure if this is the best approach. @jjsjann123 would love to know your thoughts and suggestions.

I'm not seeing use in having device as dynamic input. I haven't really seeing deivce changes across program. I think it's safe to just bake them in as-is. We probably  cc'ing @tfogal @t-vi  since they have seen a lot more models than I do.

I think we can just bake them in as constant. which I thought is already the case when they are given as inputs: https://github.com/Lightning-AI/lightning-thunder/blob/067f15aae47ad71229732ca6c35a5d190135e48c/thunder/core/jit_ext.py#L570

So I think this is yet another case where we need to have `torch/__init__.py::device` inserting a constraint.

**Comment by tfogal** [comment]
> I'm interesting in seeing how the trace is constructed with `symbolic values`
> 
> > I think one way to support this will be to make torch.device actually a symbol which can take int or IntegerProxy and return a device proxy for consumption. However, I am not sure if this is the best approach. @jjsjann123 would love to know your thoughts and suggestions.
> 
> I'm not seeing use in having device as dynamic input. I haven't really seeing deivce changes across program. I think it's safe to just bake them in as-is. We probably cc'ing @tfogal @t-vi since they have seen a lot more models than I do.

I agree; AFAICT when devs want to change devices, they take some existing thing and `to()` it to the new device. At that point, whether they invoke `device` on the pre-`to()`'d thing or the `to()`'s result, the device is effectively a constant.

Even when the device is an argument, as it is in the `foo` here, over the long run I'd expect thunder would want to do a different kind of compile for different devices. That is, you'd probably want an entirely different cost model if invoked as `jfoo("cpu", ...)` rather than `jfoo("cuda:0", ...)`

**Review by tfogal — APPROVED** [review]

**Comment by kshitij12345** [comment]
> I'm interesting in seeing how the trace is constructed with symbolic values

Snippet
```python
import thunder
import torch

def foo(dev, idx):
    return torch.ones(1, device=torch.device(dev, idx))

# jfoo = thunder.jit(foo)  # works
jfoo = thunder.jit(foo, cache=thunder.CACHE_OPTIONS.SYMBOLIC_VALUES)
print(jfoo("cuda", 0))  # fails as our current implementation can't handle `IntegerProxy`.
print(jfoo("cuda", 1))

# print(thunder.last_prologue_traces(jfoo)[-1])
# print(thunder.last_traces(jfoo)[-1])
```

Currently, the snippet fails with error -> `ValueError: [IntegerProxy name=i1, value=0] had an unexpected type <class 'thunder.core.proxies.IntegerProxy'>. Supported types are <class 'int'>` as this implementation of `torch.device` doesn't handle `IntegerProxy`.

Since, this snippet is now tracked in #463, we can fix this separately in another PR.

File: thunder/tests/test_core.py [hunk 1]
```diff
@@ -2656,3 +2656,73 @@ def foo_torch(x):
...
+            expected = foo(input)
+            assert actual.device == expected.device
+
+    # Test with str input
+    device_strs = ("cpu", "cuda", "cuda:0", "meta")
+
+    def foo1(dev):
+        # If we return the device here, thunder.jit version will return `thunder.device`
+        # while eager will return `torch.device`
+        # https://github.com/Lightning-AI/lightning-thunder/issues/573
```
Comment (mruberry) [inline thunder/tests/test_core.py:2681]: Great comment and nice link to the issue

File: thunder/torch/__init__.py [hunk 2]
```diff
@@ -4371,6 +4371,33 @@ def softmax(a: TensorLike, dim: int, dtype: None | dtypeLike = None, _stacklevel
     return _softmax(a, dim=dim, dtype=dtype)
 
 
+def torch_device(device_or_str: DeviceLike, /, index: int | None = None) -> devices.Device:
+    if isinstance(device_or_str, (devices.Device, torch.device)):
+        # PyTorch behavior:
+        # >>> torch.device(torch.device("cuda"), 0)
+        # TypeError: device(): argument 'type' (position 1) must be str, not torch.device
```
Comment (mruberry) [inline thunder/torch/__init__.py:4378]: Nice detailed handling of input types

File: thunder/torch/__init__.py [hunk 3]
```diff
@@ -4371,6 +4371,33 @@ def softmax(a: TensorLike, dim: int, dtype: None | dtypeLike = None, _stacklevel
...
+        utils.check(index is None, lambda: f"device(): `index` is only allowed when `device` is a `str`.")
+        return to_device(device_or_str)
+
+    # NOTE: device_or_str is `str`
```
Comment (mruberry) [inline thunder/torch/__init__.py:4382]: Should this function explicitly check for this?

File: thunder/torch/__init__.py [hunk 4]
```diff
@@ -4371,6 +4371,33 @@ def softmax(a: TensorLike, dim: int, dtype: None | dtypeLike = None, _stacklevel
...
+    if index is not None:
+        # PyTorch behavior:
+        # >>> torch.device("cuda:0", 0)
+        # RuntimeError: type (string) must not include an index because index was passed explicitly: cuda:0
+        has_device_idx = len(device_or_str.split(":")) > 1
+        utils.check(
+            not has_device_idx,
+            lambda: f"device string must not include an index because index was passed explicitly: {device_or_str}",
```
Comment (mruberry) [inline thunder/torch/__init__.py:4390]: Nice special case again

File: thunder/torch/__init__.py [hunk 5]
```diff
@@ -4371,6 +4371,33 @@ def softmax(a: TensorLike, dim: int, dtype: None | dtypeLike = None, _stacklevel
...
+        )
+
+    return devices.Device(device_or_str, index)
+
+
+# We don't use @torchsymbol as we don't want `torch.device()` to appear in trace as a symbol.
```
Comment (mruberry) [inline thunder/torch/__init__.py:4396]: Great comment

File: thunder/torch/__init__.py [hunk 6]
```diff
@@ -4371,6 +4371,33 @@ def softmax(a: TensorLike, dim: int, dtype: None | dtypeLike = None, _stacklevel
...
+# Because of this, we need to manually register the implementation.
+_torch_to_thunder_function_map[torch.device] = torch_device
```
Comment (mruberry) [inline thunder/torch/__init__.py:4398]: We should think about adding a layer of indirection here, like a `register_function()` -- but not for this PR

**Review by mruberry — APPROVED** [review]
Awesome! Great looking ahead to consider how symbolic values may need to be constrained when passed to `torch.device()` calls

**Comment by mruberry** [comment]
@t-vi this will probably require some help merging, looks like there are some unrelated sdpa test failures

**Pull request merged**